### PR TITLE
allow reexport on newer systems

### PR DESCRIPTION
**Pull request by MarcusCalhoun-Lopez**

### CODE REVIEW

**Comment by MarcusCalhoun-Lopez** [comment]
@cjones051073, please forgive my ignorance, but in b964c0958e9f4f1878537b867fb3aecc7c86e708, is there any reason why `-Wl,-reexport_library /usr/lib/libSystem.B.dylib` is needed instead of `-Wl,-reexport_library,/usr/lib/libSystem.B.dylib`?
As far as I can tell the latter works on all systems.

**Comment by kencu** [comment]
Chris is no longer involved in legacysupport he says, Marcus. What you are proposing looks right to me. Pinging the current maintainers. @fhgwright @mascguy

**Comment by mascguy** [comment]
> Chris is no longer involved in legacysupport he says, Marcus. What you are proposing looks right to me. Pinging the current maintainers. @fhgwright @mascguy 

I'm good with the change then.

@fhgwright ?

**Comment by fhgwright** [comment]
@mascguy 
> > Chris is no longer involved in legacysupport he says, Marcus. What you are proposing looks right to me. Pinging the current maintainers. @fhgwright @mascguy
> 
> I'm good with the change then.
> 
> @fhgwright ?

I'll take a look, but I won't have time until tomorrow.  Mostly it's a matter of testing.

It would make sense to remove cjones051073 as a reviewer, however one does that.

**Review by mascguy — APPROVED** [review]

**Comment by mascguy** [comment]
> I'll take a look, but I won't have time until tomorrow. Mostly it's a matter of testing.

No worries, not urgent by any means.

> It would make sense to remove cjones051073 as a reviewer, however one does that.

Done

**Comment by fhgwright** [comment]
I tested this on many affected systems and many unaffected ones, and it appears to work correctly, in the sense that `libMacportsLegacySystem.B.dylib` is now created on all systems.  I didn't do anything to verify that that library actually works.

I did notice that the comment on `Makefile` new line 256 (old line 258) is rendered incorrect by this change, and should be removed.

Then there's the question of whether this eliminates the need for the `Portfile` logic that begins with:
```
# The legacy-support build cannot create the LegacySupportSystem library
# directly on newer systems as the reexport link option to
# /usr/lib/libSystem.B.dylib does not work, due to the file system
# library cache added in macOS11.
```

If so, that would need to be initially conditioned on being not -devel, and then removed on the next release.  If left as is, it looks like it would gracefully fail with a warning.

**Comment by fhgwright** [comment]
I should also mention that if the only issue with 11.x+ is the option syntax, then the explanation in the Portfile comment regarding the "file system library cache" is incorrect.  However, if that explanation *is* correct, then the change here might not actually help anything.

**Comment by kencu** [comment]
Well -- a 5-second test should answer that, but presumably that is exactly why Marcus did this....to fix that.

**Comment by MarcusCalhoun-Lopez** [comment]
> Well -- a 5-second test should answer that, but presumably that is exactly why Marcus did this....to fix that.

That is correct.
Assuming this PR works as intended, the Portfile can be simplified.

With `-Wl,-reexport_library /usr/lib/libSystem.B.dylib` (*without* the comma), the shell looks for the file /usr/lib/libSystem.B.dylib, which does not exists on newer systems.
With `-Wl,-reexport_library,/usr/lib/libSystem.B.dylib` (*with* the comma), the shell never knows about /usr/lib/libSystem.B.dylib, and the linker is able to handle the mapping correctly.

```
--- Portfile.orig	2024-04-03 11:07:45
+++ Portfile	2024-04-14 18:39:38
@@ -67,13 +67,9 @@
     delete ${worksrcpath}/src/macports_legacy_atexit.c
 }
 
-set max_darwin_reexport 19
-set max_darwin_optool   20
-
 build.env-append    LD=ld \
                     "LIPO=/usr/bin/lipo" \
                     PLATFORM=${os.major} \
-                    MAX_DARWIN_REEXPORT=${max_darwin_reexport} \
                     SOCURVERSION=${release_ver} \
                     SOCOMPATVERSION=${compat_ver}
 
@@ -107,45 +103,6 @@
         } else {
             ui_debug "Copying ${f} to ${to}"
             copy ${f} ${to}
-        }
-    }
-}
-
-# The legacy-support build cannot create the LegacySupportSystem library
-# directly on newer systems as the reexport link option to
-# /usr/lib/libSystem.B.dylib does not work, due to the file system
-# library cache added in macOS11.
-#
-# Fall back to using optool here. Optool does not currently work on arm64.
-# This means that LegacySupportSystem will be missing on arm64, and that
-# universal builds of LegacySupportSystem will lack the arm64 slice
-# (which will produce a warning).
-
-if { ${os.major} > ${max_darwin_reexport}
-     && ${os.major} <= ${max_darwin_optool} } {
-     # First determine whether any non-arm64 slice will be built
-    set optool_needed no
-    foreach arch ${muniversal.architectures} {
-        if { ${arch} ne "arm64"} {
-            set optool_needed yes
-        }
-    }
-    # If any non-arm64, add optool dependency
-    if { ${optool_needed} } {
-        depends_build-append port:optool
-        depends_skip_archcheck-append optool
-    }
-    # After destroot, apply optool to any non-arm64 slice
-    post-destroot {
-        if { ${muniversal.build_arch} ne "arm64" } {
-            set legSupp   ${prefix}/lib/libMacportsLegacySupport.dylib
-            set legSystem ${prefix}/lib/libMacportsLegacySystem.B.dylib
-            if {![file exists ${destroot}${legSystem}]} {
-                copy ${destroot}${legSupp} ${destroot}${legSystem}
-                system "${prefix}/bin/optool install -c reexport -p /usr/lib/libSystem.B.dylib -t ${destroot}${legSystem}"
-            } else {
-                ui_warn "${destroot}${legSystem} already exists."
-            }
         }
     }
 }
```

**Comment by fhgwright** [comment]
@MarcusCalhoun-Lopez 
> > Well -- a 5-second test should answer that, but presumably that is exactly why Marcus did this....to fix that.
> 
> That is correct. Assuming this PR works as intended, the Portfile can be simplified.

That's what I figured, but the comment regarding "file system library cache" made me suspicious.  Of course the Portfile simplification can't happen until there's a *release* version with the fix, though it could be conditionaled out in the -devel case.

If this had happened sooner, it would have saved me the trouble of reworking the `optool` logic to handle universal builds correctly. :-)

I presume you'll repush with the obsolete comment removed.

**Comment by MarcusCalhoun-Lopez** [comment]
> I presume you'll repush with the obsolete comment removed.

Thank you for noticing.
The offending comment should now be removed in the pull request.

**Review by mascguy — APPROVED** [review]

**Comment by fhgwright** [comment]
@mascguy @MarcusCalhoun-Lopez 
OK, I've now fully tested this, both "bare" and with a temporary Portfile (both with and without disabling the `optool` hack).

TL; DR; It's good to go.

Tested on:
```
Mac OS X 10.4.11 8S165, PPC, Xcode 2.5 8M2558
Mac OS X 10.4.11 8S2167, i386, Xcode 2.5 8M2558
Mac OS X 10.4.11 8S2167, x86_64, Xcode 2.5 8M2558 *
Mac OS X 10.5.8 9L31a, PPC, Xcode 3.1.4 9M2809
Mac OS X 10.5.8 9L31a, i386, Xcode 3.1.4 9M2809
Mac OS X 10.5.8 9L31a, x86_64, Xcode 3.1.4 9M2809
Mac OS X 10.5.8 9L31a, PPC (i386 Rosetta), Xcode 3.1.4 9M2809
Mac OS X 10.6.8 10K549, i386, Xcode 3.2.6 10M2518
Mac OS X 10.6.8 10K549, x86_64, Xcode 3.2.6 10M2518
Mac OS X 10.6.8 10K549, PPC (i386 Rosetta), Xcode 3.2.6 10M2518
Mac OS X 10.7.5 11G63, x86_64, Xcode 4.6.3 4H1503
OS X 10.8.5 12F2560, x86_64, Xcode 5.1.1 5B1008
OS X 10.9.5 13F1911, x86_64, Xcode 6.2 6C131e
OS X 10.10.5 14F2511, x86_64, Xcode 7.2 7C68
OS X 10.11.6 15G22010, x86_64, Xcode 8.1 8B62
macOS 10.12.6 16G2136, x86_64, Xcode 9.2 9C40b
macOS 10.13.6 17G14042, x86_64, Xcode 10.1 10B61
macOS 10.14.6 18G9323, x86_64, Xcode 11.3.1 11C505
macOS 10.15.7 19H15, x86_64, Xcode 12.4 12D4e
macOS 11.7.10 20G1427, x86_64, Xcode 13.2.1 13C100
macOS 11.7.10 20G1427, arm64, Xcode 13.2.1 13C100
macOS 12.7.4 21H1123, x86_64, Xcode 14.2 14C18
macOS 12.7.4 21H1123, arm64, Xcode 14.2 14C18
macOS 13.6.6 22G630, arm64, Xcode 15.2 15C500b
macOS 14.4.1 23E224, arm64, Xcode 15.3 15E204a
```
`*` The *project* doesn't build on 10.4 x86_64 (unrelated to this change), but the *port* does, due to coercing the architecture to i386.

I also have a preliminary Portfile to update `legacy-support-devel` (including conditionally disabling the `optool` hack), but it can't be finalized and tested until this PR is merged, since the merge will change the commit ID here.

**Comment by mascguy** [comment]
> @mascguy @MarcusCalhoun-Lopez 
> OK, I've now fully tested this, both "bare" and with a temporary Portfile (both with and without disabling the `optool` hack).
> 
> TL; DR; It's good to go.

Awesome, thanks for testing Fred!

I'll go ahead and merge this then, so that you have a final commit hash.
